File: movie-ticket-frontend/src/services/movieService.ts
import api from "../config/api";

export interface MovieDTO {
  id: number;
  title: string;
  description: string;
  genre: string;
  duration: number;
  releaseDate: string;
  posterUrl: string;
  rating: number;
  status: string;
}

export interface CreateMovieRequest {
  title: string;
  description: string;
  genre: string;
  duration: number;
  releaseDate: string;
  posterUrl: string;
}

export const movieService = {
  getAllMovies: async (): Promise<MovieDTO[]> => {
    const response = await api.get<MovieDTO[]>("/movies");
    return response.data;
  },

  getMovieById: async (id: number): Promise<MovieDTO> => {
    const response = await api.get<MovieDTO>(`/movies/${id}`);
    return response.data;
  },

  searchMovies: async (query: string): Promise<MovieDTO[]> => {
    const response = await api.get<MovieDTO[]>("/movies/search", {
      params: { query },
    });
    return response.data;
  },

  getMoviesByGenre: async (genre: string): Promise<MovieDTO[]> => {
    const response = await api.get<MovieDTO[]>(`/movies/genre/${genre}`);
    return response.data;
  },

  getUpcomingMovies: async (): Promise<MovieDTO[]> => {
    const response = await api.get<MovieDTO[]>("/movies/upcoming");
    return response.data;
  },

  createMovie: async (data: CreateMovieRequest): Promise<MovieDTO> => {
    const response = await api.post<MovieDTO>("/movies", data);
    return response.data;
  },

  updateMovie: async (
    id: number,
    data: CreateMovieRequest
  ): Promise<MovieDTO> => {
    const response = await api.put<MovieDTO>(`/movies/${id}`, data);
    return response.data;
  },

  deleteMovie: async (id: number): Promise<void> => {
    await api.delete(`/movies/${id}`);
  },
};

File: movie-ticket-system/src/main/java/com/movieticket/service/MovieService.java
package com.movieticket.service;

import com.movieticket.dto.CreateMovieRequest;
import com.movieticket.dto.MovieDTO;
import com.movieticket.entity.Movie;
import com.movieticket.repository.MovieRepository;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class MovieService {

    @Autowired
    private MovieRepository movieRepository;

    @Autowired
    private ModelMapper modelMapper;

    public MovieDTO createMovie(CreateMovieRequest request) {
        Movie movie = Movie.builder()
                .title(request.getTitle())
                .description(request.getDescription())
                .genre(request.getGenre())
                .language(request.getLanguage())
                .releaseDate(request.getReleaseDate())
                .duration(request.getDuration())
                .posterUrl(request.getPosterUrl())
                .status("ACTIVE")
                .rating(0.0)
                .build();

        Movie savedMovie = movieRepository.save(movie);
        return modelMapper.map(savedMovie, MovieDTO.class);
    }

    public MovieDTO getMovieById(Long id) {
        Movie movie = movieRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Movie not found"));
        return modelMapper.map(movie, MovieDTO.class);
    }

    public List<MovieDTO> getAllMovies() {
        return movieRepository.findAll().stream()
                .map(movie -> modelMapper.map(movie, MovieDTO.class))
                .collect(Collectors.toList());
    }

    public List<MovieDTO> getActiveMovies() {
        return movieRepository.findByStatus("ACTIVE").stream()
                .map(movie -> modelMapper.map(movie, MovieDTO.class))
                .collect(Collectors.toList());
    }

    public List<MovieDTO> getMoviesByGenre(String genre) {
        return movieRepository.findByGenre(genre).stream()
                .map(movie -> modelMapper.map(movie, MovieDTO.class))
                .collect(Collectors.toList());
    }

    public List<MovieDTO> searchMovies(String searchTerm) {
        return movieRepository.searchMovies(searchTerm).stream()
                .map(movie -> modelMapper.map(movie, MovieDTO.class))
                .collect(Collectors.toList());
    }

    public List<MovieDTO> getUpcomingMovies() {
        return movieRepository.findByReleaseDateAfterAndStatus(LocalDate.now(), "ACTIVE")
                .stream()
                .map(movie -> modelMapper.map(movie, MovieDTO.class))
                .collect(Collectors.toList());
    }

    public MovieDTO updateMovie(Long id, CreateMovieRequest request) {
        Movie movie = movieRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Movie not found"));

        movie.setTitle(request.getTitle());
        movie.setDescription(request.getDescription());
        movie.setGenre(request.getGenre());
        movie.setLanguage(request.getLanguage());
        movie.setReleaseDate(request.getReleaseDate());
        movie.setDuration(request.getDuration());
        movie.setPosterUrl(request.getPosterUrl());

        Movie updatedMovie = movieRepository.save(movie);
        return modelMapper.map(updatedMovie, MovieDTO.class);
    }

    public void deleteMovie(Long id) {
        Movie movie = movieRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Movie not found"));
        movie.setStatus("ENDED");
        movieRepository.save(movie);
    }
}
